### PR TITLE
Optimize ascii.valid

### PR DESCRIPTION
**Pull request by achille-roussel**
Taking a similar approach to what we did for #66, applied to the `ascii.valid` function:

```
name                           old time/op    new time/op     delta
Valid/a_string_of_16B.           7.76ns ± 2%     7.65ns ± 3%      ~     (p=0.206 n=5+5)
Valid/an_invalid_string...       8.47ns ± 2%     7.38ns ± 1%   -12.91%  (p=0.008 n=5+5)
Valid/some_kind_of_long...       10.3ns ± 2%      8.1ns ± 2%   -21.01%  (p=0.008 n=5+5)
Valid/some_kind_of_long...#01    8.87ns ± 2%     5.73ns ± 4%   -35.33%  (p=0.008 n=5+5)
Valid/12345678901234567...        559ns ± 6%      127ns ± 3%   -77.24%  (p=0.008 n=5+5)

name                           old speed      new speed       delta
Valid/a_string_of_16B.         2.06GB/s ± 2%   2.09GB/s ± 3%      ~     (p=0.222 n=5+5)
Valid/an_invalid_string...     3.90GB/s ± 2%   4.47GB/s ± 1%   +14.83%  (p=0.008 n=5+5)
Valid/some_kind_of_long...     5.06GB/s ± 2%   6.42GB/s ± 2%   +26.95%  (p=0.008 n=5+5)
Valid/some_kind_of_long...#01  7.22GB/s ± 2%  11.17GB/s ± 4%   +54.67%  (p=0.008 n=5+5)
Valid/12345678901234567...     17.9GB/s ± 6%   78.6GB/s ± 3%  +338.98%  (p=0.008 n=5+5)
```

### CODE REVIEW

File: ascii/valid_print_asm.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
+// +build ignore
```
Comment (achille-roussel) [inline ascii/valid_print_asm.go:1]: This was `valid_asm.go`, containing the `validPrint16` function, moved to `valid_print_asm.go`.

File: ascii/valid_asm.go [hunk 2]
```diff
@@ -8,73 +8,40 @@ import (
...
-	VPCMPGTB(maxYMM, ymm1, ymm6)
-	VPANDN(ymm5, ymm6, ymm7)
-	VPAND(ymm4, ymm7, ymm8)      // combine masks
-	VPMOVMSKB(ymm8, msk0)
-	XORL(U32(0xFFFFFFFF), msk0)  // check for a zero somewhere
+
+	VPMOVMSKB(ymm0, msk0)
+	VPMOVMSKB(ymm1, msk1)
+	ORL(msk1, msk0)
+	CMPL(msk0, Imm(0))
```
Comment (chriso) [inline ascii/valid_asm.go:39]: This is a bit faster for me:

```golang
VMOVUPS(Mem{Base: p}, ymm0)
VPOR((Mem{Base: p}).Offset(32), ymm0, ymm1)
VPMOVMSKB(ymm1, msk0)
CMPL(msk0, Imm(0))
```

```
Valid/12345678901234567...-4          70.3GB/s ± 0%  88.4GB/s ± 0%  +25.64%  (p=0.008 n=5+5)
```

**Review by chriso — APPROVED** [review]

File: ascii/valid_asm.go (hunk 2 above)
Comment (achille-roussel) [reply ascii/valid_asm.go:39]: ah nice! I'll add that in

Comment (achille-roussel) [reply ascii/valid_asm.go:39]:
```
$ benchstat /tmp/bench{5,6}
name                           old time/op    new time/op     delta
Valid/a_string_of_16B.           7.76ns ± 2%     7.44ns ± 4%    -4.15%  (p=0.040 n=5+5)
Valid/an_invalid_string...       8.47ns ± 2%     6.94ns ± 6%   -18.11%  (p=0.008 n=5+5)
Valid/some_kind_of_long...       10.3ns ± 2%      8.4ns ± 2%   -17.86%  (p=0.008 n=5+5)
Valid/some_kind_of_long...#01    8.87ns ± 2%     5.78ns ± 3%   -34.83%  (p=0.008 n=5+5)
Valid/12345678901234567...        559ns ± 6%      103ns ± 3%   -81.50%  (p=0.008 n=5+5)

name                           old speed      new speed       delta
Valid/a_string_of_16B.         2.06GB/s ± 2%   2.15GB/s ± 4%    +4.38%  (p=0.032 n=5+5)
Valid/an_invalid_string...     3.90GB/s ± 2%   4.76GB/s ± 6%   +22.18%  (p=0.008 n=5+5)
Valid/some_kind_of_long...     5.06GB/s ± 2%   6.17GB/s ± 2%   +22.09%  (p=0.008 n=5+5)
Valid/some_kind_of_long...#01  7.22GB/s ± 2%  11.08GB/s ± 3%   +53.48%  (p=0.008 n=5+5)
Valid/12345678901234567...     17.9GB/s ± 6%   96.8GB/s ± 3%  +440.52%  (p=0.008 n=5+5)
```

🚀

Comment (achille-roussel) [reply ascii/valid_asm.go:39]: Added in https://github.com/segmentio/encoding/commit/2c5ebb4ec4eebcd65fce51dbfae445df53a6a6d7

Comment (achille-roussel) [reply ascii/valid_asm.go:39]: https://github.com/segmentio/encoding/pull/70

**Pull request merged**